syntax = "proto2";

package rv.protos;

import "rastervision/protos/machine_learning.proto";

message TrainConfig {

    message ObjectDetectionOptions {
        // path to Object Detection's train.py script
        optional string train_py = 1 [default="/opt/tf-models/object_detection/train.py"];

        // path to Object Detection's eval.py script
        optional string eval_py = 2 [default="/opt/tf-models/object_detection/eval.py"];

        // path to Object Detection's export_inference_graph.py script
        optional string export_py = 3 [default="/opt/tf-models/object_detection/export_inference_graph.py"];
    }

    message SegmentationOptions {
        optional string train_py = 1 [default="/opt/tf-models/deeplab/train.py"];
        required string tf_initial_checkpoint = 2;
        optional int32 training_number_of_steps = 3 [default=1];
        optional string train_split = 4;
        optional string model_variant = 5;
        repeated int32 atrous_rates = 6;
        optional int32 output_stride = 7 [default=16];
        optional int32 decoder_output_stride = 8 [default=4];
        repeated int32 train_crop_size = 9;
        optional int32 train_batch_size = 10 [default=1];
        optional string dataset = 11;
    }

    message Options {
        // URI for a separate training configuration that is specific to an ML
        // backend.
        optional string backend_config_uri = 1;

        // URI for the directory containing the output of make_training_chips.
        optional string training_data_uri = 2;

        // URI for the directory containing the output of training.
        optional string output_uri = 3;

        // URI for the pretrained model file.
        optional string pretrained_model_uri = 4;

        // How often to sync output of training to the cloud (in seconds).
        optional int32 sync_interval = 5 [default=600];

        oneof ml_options_type {
            ObjectDetectionOptions object_detection_options = 6;
            SegmentationOptions segmentation_options = 7;
        }

    }

    required MachineLearning machine_learning = 1;
    required Options options = 2;
}
